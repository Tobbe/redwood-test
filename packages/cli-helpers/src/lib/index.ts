import fs from 'fs'
import path from 'path'

import * as babel from '@babel/core'
import Listr from 'listr'
import { memoize } from 'lodash'
import { format } from 'prettier'

import {
  getPaths as getRedwoodPaths,
  resolveFile as internalResolveFile,
} from '@redwoodjs/internal/dist/paths'

import c from './colors'

export function isErrorWithMessage(e: any): e is { message: string } {
  return !!e.message
}

/**
 * This wraps the core version of getPaths into something that catches the exception
 * and displays a helpful error message.
 */
export const _getPaths = () => {
  try {
    return getRedwoodPaths()
  } catch (e) {
    if (isErrorWithMessage(e)) {
      console.error(c.error(e.message))
    }

    process.exit(1)
  }
}

export const getPaths = memoize(_getPaths)
export const resolveFile = internalResolveFile

export const getGraphqlPath = () => {
  return resolveFile(path.join(getPaths().api.functions, 'graphql'))
}

export const graphFunctionDoesExist = () => {
  const graphqlPath = getGraphqlPath()
  return graphqlPath && fs.existsSync(graphqlPath)
}

// TODO: Move this into `generateTemplate` when all templates have TS support
/*
 * Convert a generated TS template file into JS.
 */
export const transformTSToJS = (filename: string, content: string) => {
  const babelFileResult = babel.transform(content, {
    filename,
    // If you ran `yarn rw generate` in `./web` transformSync would import the `.babelrc.js` file,
    // in `./web`? despite us setting `configFile: false`.
    cwd: process.env.NODE_ENV === 'test' ? undefined : getPaths().base,
    configFile: false,
    plugins: [
      [
        '@babel/plugin-transform-typescript',
        {
          isTSX: true,
          allExtensions: true,
        },
      ],
    ],
    retainLines: true,
  })

  if (!babelFileResult?.code) {
    console.error(c.error(`Could not transform ${filename} to JS`))

    process.exit(1)
  }

  return prettify(filename.replace(/\.tsx?$/, '.js'), babelFileResult.code)
}

/**
 * This returns the config present in `prettier.config.js` of a Redwood project.
 */
export const prettierOptions = () => {
  try {
    return require(path.join(getPaths().base, 'prettier.config.js'))
  } catch (e) {
    return undefined
  }
}

export const prettify = (
  templateFilename: string,
  renderedTemplate: string
) => {
  // We format .js and .css templates, we need to tell prettier which parser
  // we're using.
  // https://prettier.io/docs/en/options.html#parser
  const parser = {
    '.css': 'css',
    '.js': 'babel',
    '.ts': 'babel-ts',
    '.tsx': 'babel-ts',
  }[path.extname(templateFilename.replace('.template', ''))]

  if (typeof parser === 'undefined') {
    return renderedTemplate
  }

  return format(renderedTemplate, {
    ...prettierOptions(),
    parser,
  })
}

export const writeFile = (
  target: string,
  contents: string,
  { overwriteExisting = false } = {},
  task: Listr.ListrTaskWrapper = {} as Listr.ListrTaskWrapper // TODO: Remove type cast
) => {
  const { base } = getPaths()
  task.title = `Writing \`./${path.relative(base, target)}\``
  if (!overwriteExisting && fs.existsSync(target)) {
    throw new Error(`${target} already exists.`)
  }

  const filename = path.basename(target)
  const targetDir = target.replace(filename, '')
  fs.mkdirSync(targetDir, { recursive: true })
  fs.writeFileSync(target, contents)
  task.title = `Successfully wrote file \`./${path.relative(base, target)}\``
}

/**
 * Creates a list of tasks that write files to the disk.
 *
 * @param files - {[filepath]: contents}
 */
export const writeFilesTask = (
  files: Record<string, string>,
  options: { overwriteExisting: boolean }
) => {
  const { base } = getPaths()
  return new Listr(
    Object.keys(files).map((file) => {
      const contents = files[file]

      return {
        title: `...waiting to write file \`./${path.relative(base, file)}\`...`,
        task: (_ctx: Listr.ListrContext, task: Listr.ListrTaskWrapper) => {
          return writeFile(file, contents, options, task)
        },
      }
    })
  )
}
