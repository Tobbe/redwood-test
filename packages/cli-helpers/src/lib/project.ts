import fs from 'fs'
import path from 'path'

import { getPaths, resolveFile } from './paths'

export const isTypeScriptProject = () => {
  const paths = getPaths()
  return (
    fs.existsSync(path.join(paths.web.base, 'tsconfig.json')) ||
    fs.existsSync(path.join(paths.api.base, 'tsconfig.json'))
  )
}

export const getGraphqlPath = () => {
  return resolveFile(path.join(getPaths().api.functions, 'graphql'))
}

export const graphFunctionDoesExist = () => {
  const graphqlPath = getGraphqlPath()
  return graphqlPath && fs.existsSync(graphqlPath)
}
