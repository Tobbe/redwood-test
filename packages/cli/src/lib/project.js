import fs from 'fs'
import path from 'path'

import { getPaths } from '@redwoodjs/cli-helpers'

export const sides = () => {
  const paths = getPaths()

  let sides = []
  if (fs.existsSync(path.join(paths.web.base, 'package.json'))) {
    sides = [...sides, 'web']
  }
  if (fs.existsSync(path.join(paths.api.base, 'package.json'))) {
    sides = [...sides, 'api']
  }
  return sides
}
