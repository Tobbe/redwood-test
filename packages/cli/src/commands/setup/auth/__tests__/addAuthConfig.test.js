// Have to use `var` here to avoid "Temporal Dead Zone" issues
var mockWebAppPath = ''
var mockWebRoutesPath = ''

import fs from 'fs'

import '../../../../lib/mockTelemetry'

import {
  addConfigToApp,
  addConfigToRoutes,
  createWebAuthTs,
} from '../authTasks'

jest.mock('../../../../lib', () => {
  const path = require('path')
  const __dirname = path.resolve()
  const originalModule = jest.requireActual('../../../../lib')

  return {
    transformTSToJS: originalModule.transformTSToJS,
    resolveFile: originalModule.resolveFile,
    getPaths: () => ({
      api: { functions: '', src: '', lib: '' },
      web: {
        src: path.join(__dirname, '../create-redwood-app/template/web/src'),
        app: path.join(
          __dirname,
          mockWebAppPath || '../create-redwood-app/template/web/src/App.tsx'
        ),
        routes: path.join(
          __dirname,
          mockWebRoutesPath ||
            '../create-redwood-app/template/web/src/Routes.tsx'
        ),
      },
      base: path.join(__dirname, '../create-redwood-app/template'),
    }),
  }
})

jest.mock('@redwoodjs/internal/dist/paths', () => {
  const path = require('path')
  const __dirname = path.resolve()
  const originalModule = jest.requireActual('@redwoodjs/internal/dist/paths')

  return {
    transformTSToJS: originalModule.transformTSToJS,
    resolveFile: originalModule.resolveFile,
    getPaths: () => ({
      api: { functions: '', src: '', lib: '' },
      web: {
        src: path.join(__dirname, '../create-redwood-app/template/web/src'),
        app: path.join(
          __dirname,
          mockWebAppPath || '../create-redwood-app/template/web/src/App.tsx'
        ),
        routes: path.join(
          __dirname,
          '../create-redwood-app/template/web/src/Routes.tsx'
        ),
      },
      base: path.join(__dirname, '../create-redwood-app/template'),
    }),
  }
})

jest.mock('../../../../lib/project', () => ({
  isTypeScriptProject: () => false,
}))

// This function checks output matches
const writeFileSyncSpy = jest.fn((_, content) => {
  // Line breaks cause an issue on windows snapshots
  expect(content).toMatchSnapshot()
})

beforeEach(() => {
  mockWebAppPath = ''
  jest.restoreAllMocks()
  jest.spyOn(fs, 'writeFileSync').mockImplementation(writeFileSyncSpy)
})

describe('Should update App.{js,tsx}, Routes.{js,tsx} and add auth.ts', () => {
  it('Matches Auth0 Snapshot', () => {
    addConfigToApp()
    createWebAuthTs('auth0')
    addConfigToRoutes()
  })

  it('Matches Clerk Snapshot', () => {
    addConfigToApp()
    createWebAuthTs('clerk')
    addConfigToRoutes()
  })
})

describe('Components with props', () => {
  it('Should add useAuth on the same line for single line components, and separate line for multiline components', () => {
    mockWebAppPath =
      'src/commands/setup/auth/__tests__/fixtures/AppWithCustomRedwoodApolloProvider.js'
    mockWebRoutesPath =
      'src/commands/setup/auth/__tests__/fixtures/RoutesWithCustomRouterProps.tsx'

    addConfigToApp()
    addConfigToRoutes()
  })

  it('Should not add useAuth if one already exists', () => {
    mockWebAppPath =
      'src/commands/setup/auth/__tests__/fixtures/AppWithCustomRedwoodApolloProvider.js'
    mockWebRoutesPath =
      'src/commands/setup/auth/__tests__/fixtures/RoutesWithUseAuth.tsx'

    addConfigToApp()
    addConfigToRoutes()
  })
})

describe('Customized App.js', () => {
  it('Should add auth config when using explicit return', () => {
    mockWebAppPath =
      'src/commands/setup/auth/__tests__/fixtures/AppWithExplicitReturn.js'

    addConfigToApp()
  })
})

describe('Swapped out GraphQL client', () => {
  let consoleWarn

  beforeAll(() => {
    consoleWarn = console.warn
    console.warn = jest.fn()
  })

  afterAll(() => {
    console.warn = consoleWarn
  })

  it('Should add auth config when app is missing RedwoodApolloProvider', () => {
    mockWebAppPath =
      'src/commands/setup/auth/__tests__/fixtures/AppWithoutRedwoodApolloProvider.js'

    addConfigToApp()

    expect(console.warn).toHaveBeenCalledWith(
      expect.stringMatching(/GraphQL.*useAuth/)
    )
  })
})
