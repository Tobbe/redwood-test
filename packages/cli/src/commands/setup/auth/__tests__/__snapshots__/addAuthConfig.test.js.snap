// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components with props Should add useAuth on the same line for single line components, and separate line for multiline components 1`] = `
"import { FatalErrorBoundary, RedwoodProvider } from '@redwoodjs/web'
import { RedwoodApolloProvider } from '@redwoodjs/web/apollo'

import FatalErrorPage from 'src/pages/FatalErrorPage'
import Routes from 'src/Routes'

import { AuthProvider, useAuth } from './auth'

import './index.css'

const App = () => (
  <FatalErrorBoundary page={FatalErrorPage}>
    <RedwoodProvider titleTemplate=\\"%PageTitle | %AppTitle\\">
      <AuthProvider>
        <RedwoodApolloProvider useAuth={useAuth} graphQLClientConfig={{ cache }}>
          <Routes />
        </RedwoodApolloProvider>
      </AuthProvider>
    </RedwoodProvider>
  </FatalErrorBoundary>
)

export default App
"
`;

exports[`Components with props Should add useAuth on the same line for single line components, and separate line for multiline components 2`] = `
"// In this file, all Page components from 'src/pages\` are auto-imported. Nested
// directories are supported, and should be uppercase. Each subdirectory will be
// prepended onto the component name.
//
// Examples:
//
// 'src/pages/HomePage/HomePage.js'         -> HomePage
// 'src/pages/Admin/BooksPage/BooksPage.js' -> AdminBooksPage

import { Router, Route } from '@redwoodjs/router'

import { useAuth } from './auth'

const Routes = () => {
  return (
    <Router useAuth={useAuth}
      pageLoadingDelay={400}
      trailingSlashes=\\"always\\"
      paramTypes={{
        foo: {
          match: /foo/,
          parse: (value: string) => value.split('').reverse().join(''),
        },
      }}
    >
      <Route notfound page={NotFoundPage} />
    </Router>
  )
}

export default Routes
"
`;

exports[`Components with props Should not add useAuth if one already exists 1`] = `
"import { FatalErrorBoundary, RedwoodProvider } from '@redwoodjs/web'
import { RedwoodApolloProvider } from '@redwoodjs/web/apollo'

import FatalErrorPage from 'src/pages/FatalErrorPage'
import Routes from 'src/Routes'

import { AuthProvider, useAuth } from './auth'

import './index.css'

const App = () => (
  <FatalErrorBoundary page={FatalErrorPage}>
    <RedwoodProvider titleTemplate=\\"%PageTitle | %AppTitle\\">
      <AuthProvider>
        <RedwoodApolloProvider useAuth={useAuth} graphQLClientConfig={{ cache }}>
          <Routes />
        </RedwoodApolloProvider>
      </AuthProvider>
    </RedwoodProvider>
  </FatalErrorBoundary>
)

export default App
"
`;

exports[`Components with props Should not add useAuth if one already exists 2`] = `
"// In this file, all Page components from 'src/pages\` are auto-imported. Nested
// directories are supported, and should be uppercase. Each subdirectory will be
// prepended onto the component name.
//
// Examples:
//
// 'src/pages/HomePage/HomePage.js'         -> HomePage
// 'src/pages/Admin/BooksPage/BooksPage.js' -> AdminBooksPage

import { Router, Route } from '@redwoodjs/router'

import { useAuth } from './auth'

const Routes = () => {
  return (
    <Router useAuth={useAuth}
      pageLoadingDelay={400}
      trailingSlashes=\\"always\\"
      paramTypes={{
        foo: {
          match: /foo/,
          parse: (value: string) => value.split('').reverse().join(''),
        },
      }}
      useAuth={() => ({
        loading: false,
        isAuthenticated: false
      })}
    >
      <Route notfound page={NotFoundPage} />
    </Router>
  )
}

export default Routes
"
`;

exports[`Customized App.js Should add auth config when using explicit return 1`] = `
"import { useEffect } from 'react'
import { FatalErrorBoundary, RedwoodProvider } from '@redwoodjs/web'
import { RedwoodApolloProvider } from '@redwoodjs/web/apollo'

import FatalErrorPage from 'src/pages/FatalErrorPage'
import Routes from 'src/Routes'

import { AuthProvider, useAuth } from './auth'

import './index.css'

const App = (props) => {
  const { cache } = props

  useEffect(() => {
    console.log('Running my custom useEffect hook on each render.')
  })

  return (
    <FatalErrorBoundary page={FatalErrorPage}>
      <RedwoodProvider titleTemplate=\\"%PageTitle | %AppTitle\\">
        <AuthProvider>
          <RedwoodApolloProvider useAuth={useAuth}>
            <AnotherProvider>
              <Routes />
            </AnotherProvider>
          </RedwoodApolloProvider>
        </AuthProvider>
      </RedwoodProvider>
    </FatalErrorBoundary>
  )
}

export default App
"
`;

exports[`Should update App.{js,tsx}, Routes.{js,tsx} and add auth.ts Matches Auth0 Snapshot 1`] = `
"import { FatalErrorBoundary, RedwoodProvider } from '@redwoodjs/web'
import { RedwoodApolloProvider } from '@redwoodjs/web/apollo'

import FatalErrorPage from 'src/pages/FatalErrorPage'
import Routes from 'src/Routes'

import { AuthProvider, useAuth } from './auth'

import './index.css'

const App = () => (
  <FatalErrorBoundary page={FatalErrorPage}>
    <RedwoodProvider titleTemplate=\\"%PageTitle | %AppTitle\\">
      <AuthProvider>
        <RedwoodApolloProvider useAuth={useAuth}>
          <Routes />
        </RedwoodApolloProvider>
      </AuthProvider>
    </RedwoodProvider>
  </FatalErrorBoundary>
)

export default App
"
`;

exports[`Should update App.{js,tsx}, Routes.{js,tsx} and add auth.ts Matches Auth0 Snapshot 2`] = `
"import { Auth0Client } from '@auth0/auth0-spa-js'

import { createAuth0Auth } from '@redwoodjs/auth-providers'

const auth0 = new Auth0Client({
  domain: process.env.AUTH0_DOMAIN || '',
  client_id: process.env.AUTH0_CLIENT_ID || '',
  redirect_uri: process.env.AUTH0_REDIRECT_URI,

  // ** NOTE ** Storing tokens in browser local storage provides persistence across page refreshes and browser tabs.
  // However, if an attacker can achieve running JavaScript in the SPA using a cross-site scripting (XSS) attack,
  // they can retrieve the tokens stored in local storage.
  // https://auth0.com/docs/libraries/auth0-spa-js#change-storage-options
  cacheLocation: 'localstorage',
  audience: process.env.AUTH0_AUDIENCE,

  // @MARK useRefreshTokens is required for automatically extending sessions
  // beyond that set in the initial JWT expiration.
  //
  // @see https://auth0.com/docs/tokens/refresh-tokens
  // useRefreshTokens: true,
})

export const { AuthProvider, useAuth } = createAuth0Auth(auth0)
"
`;

exports[`Should update App.{js,tsx}, Routes.{js,tsx} and add auth.ts Matches Auth0 Snapshot 3`] = `
"// In this file, all Page components from 'src/pages\` are auto-imported. Nested
// directories are supported, and should be uppercase. Each subdirectory will be
// prepended onto the component name.
//
// Examples:
//
// 'src/pages/HomePage/HomePage.js'         -> HomePage
// 'src/pages/Admin/BooksPage/BooksPage.js' -> AdminBooksPage

import { Router, Route } from '@redwoodjs/router'

import { useAuth } from './auth'

const Routes = () => {
  return (
    <Router useAuth={useAuth}>
      <Route notfound page={NotFoundPage} />
    </Router>
  )
}

export default Routes
"
`;

exports[`Should update App.{js,tsx}, Routes.{js,tsx} and add auth.ts Matches Clerk Snapshot 1`] = `
"import { FatalErrorBoundary, RedwoodProvider } from '@redwoodjs/web'
import { RedwoodApolloProvider } from '@redwoodjs/web/apollo'

import FatalErrorPage from 'src/pages/FatalErrorPage'
import Routes from 'src/Routes'

import { AuthProvider, useAuth } from './auth'

import './index.css'

const App = () => (
  <FatalErrorBoundary page={FatalErrorPage}>
    <RedwoodProvider titleTemplate=\\"%PageTitle | %AppTitle\\">
      <AuthProvider>
        <RedwoodApolloProvider useAuth={useAuth}>
          <Routes />
        </RedwoodApolloProvider>
      </AuthProvider>
    </RedwoodProvider>
  </FatalErrorBoundary>
)

export default App
"
`;

exports[`Should update App.{js,tsx}, Routes.{js,tsx} and add auth.ts Matches Clerk Snapshot 2`] = `
"import React, { useEffect } from 'react'

import { ClerkLoaded, ClerkProvider, useUser } from '@clerk/clerk-react'

import { createClerkAuth } from '@redwoodjs/auth-providers'
import { navigate } from '@redwoodjs/router'

export const { AuthProvider: ClerkRwAuthProvider, useAuth } = createClerkAuth()

const ClerkStatusUpdater = () => {
  const { isSignedIn, user, isLoaded } = useUser()
  const { reauthenticate } = useAuth()

  useEffect(() => {
    if (isLoaded) {
      reauthenticate()
    }
  }, [isSignedIn, user, reauthenticate, isLoaded])

  return null
}

export const AuthProvider = ({ children }) => {
  const frontendApi = process.env.CLERK_FRONTEND_API_URL
  if (!frontendApi) {
    throw new Error('Need to define env variable CLERK_FRONTEND_API_URL')
  }

  return (
    <ClerkProvider frontendApi={frontendApi} navigate={(to) => navigate(to)}>
      <ClerkRwAuthProvider>
        <ClerkLoaded>{children}</ClerkLoaded>

        <ClerkStatusUpdater />
      </ClerkRwAuthProvider>
    </ClerkProvider>
  )
}
"
`;

exports[`Should update App.{js,tsx}, Routes.{js,tsx} and add auth.ts Matches Clerk Snapshot 3`] = `
"// In this file, all Page components from 'src/pages\` are auto-imported. Nested
// directories are supported, and should be uppercase. Each subdirectory will be
// prepended onto the component name.
//
// Examples:
//
// 'src/pages/HomePage/HomePage.js'         -> HomePage
// 'src/pages/Admin/BooksPage/BooksPage.js' -> AdminBooksPage

import { Router, Route } from '@redwoodjs/router'

import { useAuth } from './auth'

const Routes = () => {
  return (
    <Router useAuth={useAuth}>
      <Route notfound page={NotFoundPage} />
    </Router>
  )
}

export default Routes
"
`;

exports[`Swapped out GraphQL client Should add auth config when app is missing RedwoodApolloProvider 1`] = `
"import { FatalErrorBoundary, RedwoodProvider } from '@redwoodjs/web'

import FatalErrorPage from 'src/pages/FatalErrorPage'
import Routes from 'src/Routes'

import { AuthProvider, useAuth } from './auth'

import './index.css'

const queryClient = {}

const App = () => (
  <FatalErrorBoundary page={FatalErrorPage}>
    <RedwoodProvider titleTemplate=\\"%PageTitle | %AppTitle\\">
      <AuthProvider>
        <QueryClientProvider client={queryClient}>
          <RedwoodReactQueryProvider>
            <Routes />
          </RedwoodReactQueryProvider>
        </QueryClientProvider>
      </AuthProvider>
    </RedwoodProvider>
  </FatalErrorBoundary>
)

export default App
"
`;
